crates/dbs-miniball/docs/MemoryVirtualization.md

```diff
@@ -0,0 +1,381 @@
+
+```c
+struct e820_entry {
+    uint64_t addr;	/* start of memory segment*/
+    uint64_t size;	/* size of memory segment*/
+    uint64_t type;	/* type of memory segment*/
+} __attribute__((packed));
+```
+
+QEMU/KVM 作为虚拟化核心组件，其可以用于模拟 X86 运行环境，因此其具备构建 X86 平台硬件信息的能力，QEMU/KVM 同样提供了创建 E820 的机制，用于为 seaBIOS 模拟 E820 相关的硬件信息，其中包括了 BIOS 的 CMOS 内存，hw_cfg 固件信息等。
+
+内核在初始化过程中，实时模式中从 BIOS 中获得并构建 E820 表，然后结合 CMDLINE 参数和 boot_params 参数构建内核自身的 E820 表，并基于该表构建系统的 IORESOURCE 信息和固件信息，最后基于 E820 表中可用内存区域信息构建 BootMEM 或者 MEMBLOCK 内存分配器
+
+### 代码解读
+在 Miniball 中，在启动虚拟机前进行加载内核、加载命令行和配置启动参数的操作。以上过程，通过 `load_kernel()` 完成，并返回 `load_result`，以供计算获得内核加载的地址，用于启动虚拟机。
+
+```rust
+pub fn run(&mut self) -> Result<()> {
+    let load_result = self.load_kernel()?;
+    #[cfg(target_arch = "x86_64")]
+    let kernel_load_addr = self.compute_kernel_load_addr(&load_result)?;
+
+    ...
+
+    self.vm.run(Some(kernel_load_addr)).map_err(Error::Vm)?;
+    loop {
+        match self.event_mgr.run() {
+            Ok(_) => (),
+            Err(e) => eprintln!("Failed to handle events: {:?}", e),
+        }
+        if !self.exit_handler.keep_running() {
+            break;
+        }
+    }
+    self.vm.shutdown();
+
+    Ok(())
+    }
+```
+
+加载内核的过程如下：
+
+1. 加载内核镜像到 guest 内存中
+1. 加载内核命令行到 guest 内存中
+1. 配置启动参数
+
+#### 加载内核镜像
+第 4 行，加载了 kernel 的镜像。第 6 ～ 24 行，通过 ELF 将 kernel 的镜像加载到 Guest 内存中。
+
+```rust
+// Load the kernel into guest memory.
+#[cfg(target_arch = "x86_64")]
+fn load_kernel(&mut self) -> Result<KernelLoaderResult> {
+    let mut kernel_image = File::open(&self.kernel_cfg.path).map_err(Error::IO)?;
+
+    // Load the kernel into guest memory.
+    let kernel_load = match Elf::load(
+        &self.guest_memory,
+        None,
+        &mut kernel_image,
+        Some(GuestAddress(self.kernel_cfg.load_addr)),
+    ) {
+        Ok(result) => result,
+        Err(loader::Error::Elf(elf::Error::InvalidElfMagicNumber)) => BzImage::load(
+            &self.guest_memory,
+            None,
+            &mut kernel_image,
+            Some(GuestAddress(self.kernel_cfg.load_addr)),
+        )
+        .map_err(Error::KernelLoad)?,
+        Err(e) => {
+            return Err(Error::KernelLoad(e));
+        }
+    };
+
+    ...
+    
+    Ok(kernel_load)
+}
+```
+
+#### 加载内核命令行
+第 6 ～ 13 行，加载内核命令行到 Guest 内存中，传入 Guest memory 和命令行起始地址，并保留命令行地址待之后输入到启动参数中。
+
+```rust
+// Load the kernel into guest memory.
+#[cfg(target_arch = "x86_64")]
+fn load_kernel(&mut self) -> Result<KernelLoaderResult> {
+    ...
+
+    // Load the kernel command line into guest memory.
+    let cmdline_addr = dbs_boot::layout::CMDLINE_START;
+    linux_loader::loader::load_cmdline(
+        &self.guest_memory,
+        GuestAddress(cmdline_addr),
+        &self.kernel_cfg.cmdline,
+    )
+    .map_err(Error::LoadCommandline)?;
+    
+    ...
+}
+```
+
+#### 配置启动参数
+第 6 ～ 16 行，配置启动参数，参数的详细内容在之后代码中介绍。
```

Comment:
Where is the detailed introduction?

Reply:
added

Reply:
Sorry, I am still implementing virtio net in Miniball, and I will supplement the description in this part later.

Reply:
Sorry, I am still implementing virtio net in Miniball, and I will supplement the description in this part after I implementing virtio net.